pkg/rpc/server/server.go

```diff
@@ -270,17 +270,39 @@ func (s *Server) handleHTTPRequest(w http.ResponseWriter, httpRequest *http.Requ
 		return
 	}
 
-	err := req.DecodeData(httpRequest.Body)
+	err := in.DecodeData(httpRequest.Body)
 	if err != nil {
-		s.writeHTTPErrorResponse(req, w, response.NewParseError("Problem parsing JSON-RPC request body", err))
+		s.writeHTTPErrorResponse(request.NewRequest(), w, response.NewParseError("Problem parsing JSON-RPC request body", err))
 		return
 	}
 
-	resp := s.handleRequest(req, nil)
-	s.writeHTTPServerResponse(req, w, resp)
+	resp := s.handleIn(in, nil)
+	s.writeHTTPServerResponse(in, w, resp)
 }
 
-func (s *Server) handleRequest(req *request.In, sub *subscriber) response.Abstract {
+func (s *Server) handleIn(in *request.In, sub *subscriber) interface{} {
+	if in.Request != nil {
+		return s.handleRequest(in.Request, sub)
+	}
+	resp := make([]response.Abstract, len(in.Batch))
+	ch := make(chan struct{})
+	for i, req := range in.Batch {
+		go func(i int, req *request.Request) {
+			if req.Error != nil {
+				resp[i] = s.packResponse(req, nil, response.NewParseError("Problem parsing JSON-RPC request body", req.Error))
+			} else {
+				resp[i] = s.handleRequest(req, sub)
+			}
+			ch <- struct{}{}
+		}(i, req)
+	}
+	for range in.Batch {
+		<-ch
+	}
+	return resp
+}
+
+func (s *Server) handleRequest(req *request.Request, sub *subscriber) response.Abstract {
```

Comment:
This one probably could stay as is, no changes required.

Reply:
Only the name has been changed here (`handleRequest` -> `handleIn`). I can leave it untouched, but in this case we should think about the new name for `handleRequest`, which actually performs handling of `request.Request` structure.